jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java

```diff
@@ -557,33 +555,35 @@ private void makeUnavailable(final Throwable e)
     private synchronized void initServlet()
         throws ServletException
     {
+        Servlet servlet = _servlet;
         try
         {
-            if (_servlet == null)
-                _servlet = getInstance();
-            if (_servlet == null)
-                _servlet = newInstance();
+            if (servlet == null || servlet instanceof UnavailableServlet)
+                servlet = getInstance();
+            if (servlet == null)
+                servlet = newInstance();
+            if (servlet instanceof javax.servlet.SingleThreadModel)
+            {
+                destroyInstance(servlet);
```

Comment:
Actually, this isn't right:  we should _not_ call destroyInstance() because this will call servlet.destroy(), but at this stage we will _not_ have called servlet.init().

Reply:
This has opened a rats nest  of issues!   It appears  that we do call predestroy on instances that have been passed in.  As we have tests for this potentially wrong behaviour, I think I will not fully address this in 9, but review during the merge to 10.

For now, I have split out a predestroy method, that is called on instances that have not yet been fully initialisd.